File: src/main/java/com/cloudant/cloujeau/IndexManagerService.java
package com.cloudant.cloujeau;

import static com.cloudant.cloujeau.OtpUtils.*;
import static com.cloudant.cloujeau.OtpUtils.asBinary;
import static com.cloudant.cloujeau.OtpUtils.asString;
import static com.cloudant.cloujeau.OtpUtils.tuple;

import java.io.File;
import java.io.IOException;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Map.Entry;

import org.apache.log4j.Logger;
import org.apache.lucene.analysis.Analyzer;
import org.apache.lucene.index.IndexWriter;
import org.apache.lucene.index.IndexWriterConfig;
import org.apache.lucene.queryparser.classic.QueryParser;
import org.apache.lucene.store.Directory;
import org.apache.lucene.store.LockFactory;

import com.codahale.metrics.Counter;
import com.codahale.metrics.Timer;
import com.ericsson.otp.erlang.OtpErlangAtom;
import com.ericsson.otp.erlang.OtpErlangBinary;
import com.ericsson.otp.erlang.OtpErlangObject;
import com.ericsson.otp.erlang.OtpErlangPid;
import com.ericsson.otp.erlang.OtpErlangTuple;

public class IndexManagerService extends Service {

    private class LRU {

        private int initialCapacity = 100;
        private float loadFactor = 0.75f;

        private class InnerLRU extends LinkedHashMap<String, OtpErlangPid> {
            private InnerLRU(final int initialCapacity, final float loadFactor) {
                super(initialCapacity, loadFactor, true);
            }

            @Override
            protected boolean removeEldestEntry(final Entry<String, OtpErlangPid> eldest) {
                final boolean result = size() > state.config.getInt("clouseau.max_indexes_open", 100);
                if (result) {
                    send(eldest.getValue(), tuple(atom("close"), atom("lru")));
                }
                return result;
            }
        }

        private Counter lruMisses = state.metricRegistry
                .counter("com.cloudant.clouseau:type=IndexManagerService,name=lru.misses");
        private Counter lruEvictions = state.metricRegistry
                .counter("com.cloudant.clouseau:type=IndexManagerService,name=lru.evictions");

        private Map<String, OtpErlangPid> pathToPid = new InnerLRU(initialCapacity, loadFactor);
        private Map<OtpErlangPid, String> pidToPath = new HashMap<OtpErlangPid, String>(initialCapacity, loadFactor);

        private OtpErlangPid get(final String path) {
            final OtpErlangPid pid = pathToPid.get(path);
            if (pid != null) {
                lruMisses.inc();
            }
            return pid;
        }

        private void put(final String path, final OtpErlangPid pid) {
            final OtpErlangPid prev = pathToPid.put(path, pid);
            pidToPath.remove(prev);
            pidToPath.put(pid, path);
        }

        private void remove(final OtpErlangPid pid) {
            final String path = pidToPath.remove(pid);
            pathToPid.remove(path);
            if (path != null) {
                lruEvictions.inc();
            }
        }

        private boolean isEmpty() {
            return pidToPath.isEmpty();
        }

        private void close() {
            pidToPath.forEach((pid, path) -> {
                send(pid, tuple(atom("close"), atom("closing")));
            });
        }

        private void closeByPath(final String pathPrefix) {
            pidToPath.forEach((pid, path) -> {
                if (path.startsWith(pathPrefix)) {
                    logger.info("closing lru for " + path);
                    send(pid, tuple(atom("close"), atom("closing")));
                }
            });
        }

    }

    private static final Logger logger = Logger.getLogger("clouseau.main");

    private final Timer openTimer;
    private final LRU lru;

    public IndexManagerService(final ServerState state) {
        super(state, "main");
        openTimer = state.metricRegistry.timer("com.cloudant.clouseau:type=IndexManagerService,name=opens");
        lru = new LRU();
    }

    @Override
    public OtpErlangObject handleCall(final OtpErlangTuple from, final OtpErlangObject request) throws Exception {
        if (request instanceof OtpErlangTuple) {
            final OtpErlangTuple tuple = (OtpErlangTuple) request;
            switch (asString(tuple.elementAt(0))) {
            case "open":
                return open(from, tuple);
            case "disk_size":
                return getDiskSize(asString(tuple.elementAt(1)));
            }
        }
        if (request instanceof OtpErlangAtom) {
            switch (asString(request)) {
            case "get_root_dir":
                return tuple(atom("ok"), asBinary(rootDir().getAbsolutePath()));

            case "version":
                return tuple(atom("ok"), asBinary(getClass().getPackage().getImplementationVersion()));
            }
        }
        return null;
    }

    @Override
    public void handleInfo(final OtpErlangObject request) throws Exception {
        if (request instanceof OtpErlangTuple) {
            final OtpErlangTuple tuple = (OtpErlangTuple) request;
            switch (asString(tuple.elementAt(0))) {
            case "touch_lru":
                lru.get(asString(tuple.elementAt(1)));
                break;
            }
        }
    }

    private OtpErlangObject open(final OtpErlangTuple from, final OtpErlangTuple request) throws Exception {
        final OtpErlangPid peer = (OtpErlangPid) request.elementAt(1);
        final OtpErlangBinary path = (OtpErlangBinary) request.elementAt(2);
        final OtpErlangObject analyzerConfig = request.elementAt(3);
        final String strPath = asString(path);

        final OtpErlangPid pid = lru.get(strPath);
        if (pid != null) {
            return tuple(atom("ok"), pid);
        }

        logger.info(String.format("Opening index at %s", strPath));
        try {
            return openTimer.time(() -> {
                final Analyzer analyzer = SupportedAnalyzers.createAnalyzer(analyzerConfig);
                final IndexWriter writer = newWriter(path, analyzer);
                final QueryParser qp = new ClouseauQueryParser(LuceneUtils.VERSION, "default", analyzer);
                final IndexService index = new IndexService(state, strPath, writer, qp);
                state.serviceRegistry.register(index);
                index.link(peer);
                lru.put(strPath, index.self());
                return tuple(atom("ok"), index.self());
            });
        } catch (final IOException e) {
            return tuple(atom("error"), asBinary(e.getMessage()));
        }
    }

    private OtpErlangObject getDiskSize(final String path) {
        final File indexDir = new File(rootDir(), path);
        final String[] files = indexDir.list();
        long diskSize = 0;
        if (files != null) {
            for (final String file : files) {
                diskSize += new File(indexDir, file).length();
            }
        }
        return tuple(atom("ok"), asList(tuple(atom("disk_size"), asOtp(diskSize))));
    }

    private File rootDir() {
        return new File(state.config.getString("clouseau.dir", "target/indexes"));
    }

    private IndexWriter newWriter(final OtpErlangBinary path, final Analyzer analyzer) throws Exception {
        final Directory dir = newDirectory(new File(rootDir(), asString(path)));
        final IndexWriterConfig writerConfig = new IndexWriterConfig(LuceneUtils.VERSION, analyzer);

        return new IndexWriter(dir, writerConfig);
    }

    private Directory newDirectory(final File path) throws ReflectiveOperationException {
        final String lockClassName = state.config
                .getString("clouseau.lock_class", "org.apache.lucene.store.NativeFSLockFactory");
        final Class<?> lockClass = Class.forName(lockClassName);
        final LockFactory lockFactory = (LockFactory) lockClass.getDeclaredConstructor().newInstance();

        final String dirClassName = state.config
                .getString("clouseau.dir_class", "org.apache.lucene.store.NIOFSDirectory");
        final Class<?> dirClass = Class.forName(dirClassName);
        return (Directory) dirClass.getDeclaredConstructor(File.class, LockFactory.class)
                .newInstance(path, lockFactory);
    }

}

File: clouseau/src/main/scala/com/cloudant/ziose/clouseau/IndexManagerService.scala
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy of
// the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations under
// the License.

package com.cloudant.ziose.clouseau

import java.io.File
import java.io.IOException
import java.time.{ Duration, Instant }
import java.time.temporal.ChronoUnit
import java.util.HashMap
import java.util.LinkedHashMap
import java.util.concurrent.ConcurrentHashMap
import java.util.{ Map => JMap }
import scala.collection.mutable.Map
import _root_.com.cloudant.ziose.scalang

import scalang._

import scala.collection.JavaConverters._
import java.util.HashSet
import com.cloudant.ziose.core.ProcessContext
import com.cloudant.ziose.core.Codec
import zio.ZIO

class IndexManagerService(ctx: ServiceContext[ConfigurationArgs])(implicit adapter: Adapter[_, _]) extends Service(ctx) with Instrumented {

  class LRU(initialCapacity: Int = 100, loadFactor: Float = 0.75f, trackIndexAccesses: Boolean = false) {

    class InnerLRU(initialCapacity: Int, loadFactor: Float) extends LinkedHashMap[String, Pid](initialCapacity, loadFactor, true)

    val capacity = ctx.args.config.getInt("clouseau.max_indexes_open", 100)
    val lruMisses = metrics.counter("lru.misses")
    val lruEvictions = metrics.counter("lru.evictions")

    val pathToPid: JMap[String, Pid] = new InnerLRU(initialCapacity, loadFactor)
    val pidToPath: JMap[Pid, String] = new HashMap(initialCapacity, loadFactor)
    val indexesSeen: Option[JMap[String, Long]] =
      if (trackIndexAccesses) Some(new ConcurrentHashMap(initialCapacity, loadFactor))
      else None

    def get(path: String): Pid = {
      assert(pathToPid.size == pidToPath.size)
      val pid: Pid = pathToPid.get(path)
      if (!Option(pid).isDefined) {
        lruMisses += 1
      }
      pid
    }

    def put(path: String, pid: Pid) = {
      assert(pathToPid.size == pidToPath.size)
      enforceCapacity
      val prev = pathToPid.put(path, pid)
      pidToPath.remove(prev)
      pidToPath.put(pid, path)
      trackIndexesSeen(path)
    }

    def remove(pid: Pid) = {
      assert(pathToPid.size == pidToPath.size)
      val path = pidToPath.remove(pid)
      pathToPid.remove(path)
      if (Option(path).isDefined) {
        lruEvictions += 1
      }
    }

    def isEmpty: Boolean = {
      pidToPath.isEmpty
    }

    def close() = {
      pidToPath.asScala foreach {
        kv => kv._1 ! ('close, 'closing)
      }
    }

    def closeByPath(path: String) = {
      pidToPath.asScala foreach {
        kv =>
          if (kv._2.startsWith(path)) {
            logger.info("closing lru for " + path)
            kv._1 ! ('close, 'closing)
          }
      }
    }

    private val trackIndexesSeenRange = Duration.ofDays(7).toSeconds

    private def trackIndexesSeen(path: String) = {
      indexesSeen.map({ store =>
        val cutOff = now - trackIndexesSeenRange
        store.asScala.retain { case (_, timestamp) => timestamp >= cutOff }
        store.put(path, now)
      })
    }

    def numberOfIndexesSeenRecently(duration: Long) = indexesSeen match {
      case Some(store) =>
        val reference = now - duration
        store.asScala.count { case (_, timestamp) => timestamp >= reference }
      case None => 0
    }

    private def enforceCapacity() {
      var excess = pathToPid.size - capacity
      if (excess > 0) {
        val it = pathToPid.entrySet.iterator
        while (excess > 0 && it.hasNext) {
          val eldest = it.next
          eldest.getValue ! ('close, 'lru)
          excess -= 1
        }
      }
    }

    private def now: Long = ChronoUnit.SECONDS.between(Instant.EPOCH, Instant.now)

  }

  val logger = LoggerFactory.getLogger("clouseau.main")
  val rootDir = new File(ctx.args.config.getString("clouseau.dir", "target/indexes"))
  val openTimer = metrics.timer("opens")
  val trackIndexATimes = ctx.args.config.getBoolean("clouseau.track_index_atimes", false)
  val lru = new LRU(trackIndexAccesses = trackIndexATimes)
  val waiters = Map[String, List[(Pid, Any)]]()
  val countLocksEnabled = ctx.args.config.getBoolean("clouseau.count_locks", false)
  if (countLocksEnabled) {
    val lockClass = Class.forName("org.apache.lucene.store.NativeFSLock")
    val field = lockClass.getDeclaredField("LOCK_HELD")
    field.setAccessible(true)
    val LOCK_HELD = field.get(null).asInstanceOf[HashSet[String]]
    metrics.gauge("NativeFSLock.count")(getNativeFSLockHeldSize(LOCK_HELD.asScala))
  }

  if (trackIndexATimes) {
    val indexRecencyTimes = List(
      ("1m", Duration.ofMinutes(1)),
      ("1h", Duration.ofHours(1)),
      ("1d", Duration.ofDays(1)),
      ("1w", Duration.ofDays(7))
    )

    for ((label, time) <- indexRecencyTimes) {
      metrics.gauge(s"indexes.seen.${label}")(lru.numberOfIndexesSeenRecently(time.toSeconds))
    }
  }

  def getNativeFSLockHeldSize(lockHeld: scala.collection.mutable.Set[String]) = lockHeld.synchronized {
    lockHeld.size
  }

  override def handleInit(): Unit = {
    logger.debug(s"handleInit(capacity = ${adapter.capacity})")
  }

  override def onTermination[PContext <: ProcessContext](reason: Codec.ETerm, ctx: PContext) = {
    ZIO.logTrace("onTermination")
  }

  override def handleCall(tag: (Pid, Any), msg: Any): Any = msg match {
    case OpenIndexMsg(peer: Pid, path: String, options: AnalyzerOptions) =>
      lru.get(path) match {
        case null =>
          waiters.get(path) match {
            case None =>
              val manager = self
              node.spawn(_ => {
                openTimer.time {
                  IndexService.start(node, ctx.args.config, path, options) match {
                    case ('ok, pid: Pid) =>
                      manager ! ('open_ok, path, peer, pid)
                    case error =>
                      manager ! ('open_error, path, error)
                  }
                }
              })
              waiters.put(path, List(tag))
            case Some(list) =>
              waiters.put(path, tag :: list)
          }
          'noreply
        case pid =>
          ('ok, pid)
      }
    case ('get_root_dir) =>
      ('ok, rootDir.getAbsolutePath())
    case DeleteDocMsg(path: String) =>
      lru.get(path) match {
        case null =>
          ('error, 'not_found)
        case pid: Pid =>
          pid ! 'delete
          'ok
      }
    case DiskSizeMsg(path: String) =>
      getDiskSize(path)
    case 'close_lru =>
      lru.close()
      'ok
    case CloseLRUByPathMsg(path: String) =>
      lru.closeByPath(path)
      'ok
    case 'version =>
      ('ok, getClass.getPackage.getImplementationVersion)
    case ('create_snapshot, indexName: String, snapshotDir: String) =>
      lru.get(indexName) match {
        case null =>
          createSnapshot(indexName, snapshotDir)
        case pid: Pid =>
          call(pid, ('create_snapshot, snapshotDir))
      }
  }

  override def handleInfo(msg: Any) = msg match {
    case ('open_ok, path: String, peer: Pid, pid: Pid) =>
      lru.put(path, pid)
      monitor(pid)
      node.link(peer, pid)
      replyAll(path, ('ok, pid))
      'noreply
    case ('open_error, path: String, error: Any) =>
      replyAll(path, error)
      'noreply
    case ('touch_lru, path: String) =>
      lru.get(path)
      'noreply
  }

  override def trapMonitorExit(monitored: Any, ref: Reference, reason: Any) = monitored match {
    case pid: Pid =>
      lru.remove(pid)
    case _ =>
      'ignored
  }

  private def getDiskSize(path: String) = {
    val indexDir = new File(rootDir, path)
    val files = indexDir.list()
    if (files != null) {
      val size = files.foldLeft(0L)((acc, fileName) =>
        acc + (new File(indexDir, fileName)).length())
      ('ok, List(('disk_size, size)))
    } else {
      ('ok, List(('disk_size, 0)))
    }
  }

  private def createSnapshot(indexName: String, snapshotDir: String): Any = {
    val originDir = new File(rootDir, indexName)
    // As the index is closed, we snapshot every file.
    val files = originDir.list
    try {
      ExternalSnapshotDeletionPolicy.snapshot(originDir, new File(snapshotDir), files.toSeq.asJavaCollection)
      'ok
    } catch {
      case e: IllegalStateException =>
        ('error, e.getMessage)
      case e: IOException =>
        ('error, e.getMessage)
    }
  }

  private def replyAll(path: String, msg: Any) = {
    waiters.remove(path) match {
      case Some(list) =>
        for (tag <- list) {
          Service.reply(tag, msg)
        }
      case None =>
        'ok
    }
  }

}
